go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go

```diff
@@ -32,6 +48,818 @@ func (c *Controller) processNextANPWorkItem(wg *sync.WaitGroup) bool {
+		)
+		acls = append(acls, acl)
+		return acls, nil
+	}
+
+	// we have ports defined in this rule, so its one ACL per port
+	// TODO(tssurya): There are some improvement PRs around this, adapt them in the future iterations
+	for i, port := range rule.ports {
+		extIDs = getANPRuleACLDbIDs(anpName, rule.gressPrefix, fmt.Sprintf("%d.%d", rule.priority, i),
+			controllerName, libovsdbops.ACLAdminNetworkPolicy).GetExternalIDs()
+		l4Match := constructMatchFromPorts(port)
+		match = fmt.Sprintf("%s && %s && %s", l3Match, lportMatch, l4Match)
+		acl := libovsdbops.BuildACL(
+			getANPGressPolicyACLName(anpName, rule.gressPrefix, fmt.Sprintf("%d.%d", rule.priority, i)),
+			direction,
+			int(rule.priority),
+			match,
+			getACLActionForANPRule(rule.action),
+			// TODO(tssurya): Hardcoding logging related parameters for now.
+			// Will fix this in future PRs when I add support for ANP-Logging
+			types.OvnACLLoggingMeter,
+			nbdb.ACLSeverityDebug,
+			false,
+			extIDs,
+			options,
+			types.DefaultANPACLTier,
+		)
+		acls = append(acls, acl)
+	}
+
+	return acls, nil
+}
+
+// createASForPeers creates the address-set for all the provided peers.
+// NOTE: We call ensureAS function using externalID match, so if the address-set already exists its no-op
+// This function also takes care of populating the adminNetworkPolicyPeer.namespaces cache
+// It also adds up all the podIPs that are supposed to be present in the created AS and returns them on
+// a per-rule basis so that the actual ops to transact these into the AS can be constructed using that
+func (c *Controller) createASForPeers(peers []*adminNetworkPolicyPeer, priority int32, gressPrefix,
+	anpName string, idType *libovsdbops.ObjectIDsType) (addressset.AddressSet, sets.Set[string], error) {
+
+	var addrSet addressset.AddressSet
+	var err error
+	peerPodIPs := sets.Set[string]{}
+	for _, peer := range peers {
+		// TODO: Double check how empty selector means all labels match works
+		namespaces, err := c.anpNamespaceLister.List(peer.namespaceSelector)
+		if err != nil {
+			return nil, nil, err
+		}
+		namespaceCache := make(map[string]map[string]bool)
+		// NOTE: Multiple peers may match on same podIP which is fine, we use sets to store them to avoid duplication
+		for _, namespace := range namespaces {
+			podCache, ok := namespaceCache[namespace.Name]
+			if !ok {
+				podCache = make(map[string]bool)
+				namespaceCache[namespace.Name] = podCache
+			}
+			podNamespaceLister := c.anpPodLister.Pods(namespace.Name)
+			pods, err := podNamespaceLister.List(peer.podSelector)
+			if err != nil {
+				return nil, nil, err
+			}
+			for _, pod := range pods {
+				// we don't handle HostNetworked or completed pods; unscheduled pods shall be handled via pod update path
+				if util.PodWantsHostNetwork(pod) || util.PodCompleted(pod) || !util.PodScheduled(pod) {
+					continue
+				}
+				podIPs, err := util.GetPodIPsOfNetwork(pod, &util.DefaultNetInfo{})
+				if err != nil {
```

Comment:
yeah i did think of this.. but to be more efficient, the current pod controller is stingy :D it only adds the key to the queue under very specific conditions and podIP notfound could be due to a lot of reasons so doesn't mean it will generate an update per say.. but i do see your concern (i had it too) around how each podIP fetch failure will cause ANP add failure. Technically speaking we should get an event when the pod get's annotated...but i saw the LSP not existing still and causing failures in the `getPortsFromSubject` code path and had to be more aggressive here by returning error for leveraging the retries...

Reply:
actually in my pod update watcher I do:
```
	oldPodIPs, _ := util.GetPodIPsOfNetwork(oldPod, &util.DefaultNetInfo{})
	newPodIPs, _ := util.GetPodIPsOfNetwork(newPod, &util.DefaultNetInfo{})
```
So any podIP changes should come as update, I can skip the ip fetch failure for podIPs not found here; this is specially important in IC=true case (https://github.com/ovn-org/ovn-kubernetes/pull/3655) since each zone will be processing a pod add versus object add differently. Any changes to podIPs will be handled in the pod update path.